### PR TITLE
feat(ts/analyzer): Assign Lvalue Rvalue

### PR DESCRIPTION
**Pull request by sunrabbit123**
**Description:**

When attempting an assignment, if there are duplicate values,
The value of the LHS is separated by an intersection
The value of the RHS is separated by Union.

**Related issue (if exists):**
#437 #422

### CODE REVIEW

**Comment by sunrabbit123** [comment]
Perhaps the following tests will fail.

```
failures:
    conformance::expressions::typeGuards::typeGuardNarrowsPrimitiveIntersection.ts
    conformance::types::literal::numericStringLiteralTypes.ts
    conformance::types::objectTypeLiteral::methodSignatures::methodSignaturesWithOverloads2.ts
```

first. typeGuardNarrowsPrimitiveIntersection.ts

The reason for the failure is that the priority of the intersection is not configured
This is because `(string & Tag2)` is evaluated as never.

second. numericStringLiteralTypes.ts
Its panic comes from the panic sung by IndexedAccessType.

third. methodSignatures::methodSignaturesWithOverloads2.ts
Because the law of factorization has not been done.
[Related data](https://sunrabbit.tistory.com/2)

**Comment by sunrabbit123** [comment]
1st and 3rd problem is already resolve #422

File: crates/stc_ts_type_checker/tests/conformance/types/literal/numericStringLiteralTypes.ts.stats.rust-debug [hunk 1]
```diff
@@ -2,5 +2,5 @@ Stats {
     required_error: 0,
     matched_error: 0,
     extra_error: 0,
-    panic: 0,
+    panic: 1,
```
Comment (kdy1) [inline crates/stc_ts_type_checker/tests/conformance/types/literal/numericStringLiteralTypes.ts.stats.rust-debug:5]: Is this PR ready for review?

File: crates/stc_ts_type_checker/tests/conformance/types/localTypes/localTypes4.ts.stats.rust-debug [hunk 2]
```diff
@@ -1,6 +1,6 @@
 Stats {
     required_error: 2,
     matched_error: 2,
-    extra_error: 0,
+    extra_error: 1,
```
Comment (kdy1) [inline crates/stc_ts_type_checker/tests/conformance/types/localTypes/localTypes4.ts.stats.rust-debug:4]: This should be fixed

File: crates/stc_ts_type_checker/tests/conformance/types/mapped/mappedTypeConstraints2.ts.stats.rust-debug [hunk 3]
```diff
@@ -1,6 +1,6 @@
 Stats {
-    required_error: 1,
-    matched_error: 2,
-    extra_error: 2,
-    panic: 0,
+    required_error: 3,
+    matched_error: 0,
+    extra_error: 0,
+    panic: 1,
```
Comment (kdy1) [inline crates/stc_ts_type_checker/tests/conformance/types/mapped/mappedTypeConstraints2.ts.stats.rust-debug:5]: This should be fixed

File: crates/stc_ts_type_checker/tests/conformance/types/mapped/mappedTypeRelationships.ts.stats.rust-debug [hunk 4]
```diff
@@ -1,6 +1,6 @@
 Stats {
-    required_error: 27,
-    matched_error: 1,
-    extra_error: 21,
-    panic: 0,
+    required_error: 28,
+    matched_error: 0,
+    extra_error: 0,
+    panic: 1,
```
Comment (kdy1) [inline crates/stc_ts_type_checker/tests/conformance/types/mapped/mappedTypeRelationships.ts.stats.rust-debug:5]: This should be fixed

File: crates/stc_ts_type_checker/tests/conformance/types/mapped/recursiveMappedTypes.ts.stats.rust-debug [hunk 5]
```diff
@@ -2,5 +2,5 @@ Stats {
     required_error: 9,
     matched_error: 0,
     extra_error: 0,
-    panic: 0,
+    panic: 1,
```
Comment (kdy1) [inline crates/stc_ts_type_checker/tests/conformance/types/mapped/recursiveMappedTypes.ts.stats.rust-debug:5]: This should be fixed

File: crates/stc_ts_type_checker/tests/conformance/types/objectTypeLiteral/methodSignatures/methodSignaturesWithOverloads2.ts.stats.rust-debug [hunk 6]
```diff
@@ -1,6 +1,6 @@
 Stats {
     required_error: 0,
     matched_error: 0,
-    extra_error: 0,
+    extra_error: 2,
```
Comment (kdy1) [inline crates/stc_ts_type_checker/tests/conformance/types/objectTypeLiteral/methodSignatures/methodSignaturesWithOverloads2.ts.stats.rust-debug:4]: This should be fixed

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thanks!

**Comment by sunrabbit123** [comment]
@kdy1 Thank you for your PR review
but, All the added panics originate from the following code.

```rs
 if let Type::IndexedAccessType(prop_ty) = prop_ty.normalize() {
    match prop_ty.index_type.normalize() {
        // if you want remove this, first you should overflow narrow any
        Type::Param(..) => {}
        _ => {
            panic!("{:?}", prop_ty);
        }
    }
}
```

If we don't have this code, `recursiveMappedTypes.ts` call overflow
Our code is not yet ready to remove it.

File: crates/stc_ts_file_analyzer/src/analyzer/types/mod.rs [hunk 7]
```diff
@@ -527,6 +527,27 @@ impl Analyzer<'_, '_> {
...
+                                            | TsKeywordTypeKind::TsStringKeyword
+                                            | TsKeywordTypeKind::TsSymbolKeyword,
+                                        ..
+                                    })
+                                    | Type::Lit(LitType {
+                                        lit: RTsLit::Str(..) | RTsLit::Number(..),
+                                        ..
+                                    }) => {}
+                                    _ => {
+                                        panic!("{:?}", prop_ty);
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/types/mod.rs:546]: I'll not merge this PR while this is fixed

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/stc_ts_file_analyzer/src/analyzer/types/mod.rs (hunk 7 above)
Comment (sunrabbit123) [reply crates/stc_ts_file_analyzer/src/analyzer/types/mod.rs:546]: May I ask what you are concerned about?

Comment (kdy1) [reply crates/stc_ts_file_analyzer/src/analyzer/types/mod.rs:658]: It's simply a regression

File: crates/stc_ts_type_checker/tests/conformance/types/localTypes/localTypes4.ts.stats.rust-debug (hunk 2 above)
Comment (sunrabbit123) [reply crates/stc_ts_type_checker/tests/conformance/types/localTypes/localTypes4.ts.stats.rust-debug:4]:
```ts
function f4() {
    // Local types are block scoped
    interface T { x: number }
    let v: T;
    v.x = 10;
    if (true) {
        interface T { x: string }
        let v: T;
        v.x = "hello"; // extra_error
    }
    else {
        v.x = 20;
    }
}
```
For this to be fixed, the scope must be separated.

Comment (kdy1) [reply crates/stc_ts_type_checker/tests/conformance/types/localTypes/localTypes4.ts.stats.rust-debug:4]: If so, let's postpone this. Thank you!

**Comment by sunrabbit123** [comment]
It just so happened that Readonly also resolve.

File: crates/stc_ts_type_checker/tests/conformance/controlFlow/controlFlowBindingElement.ts.stats.rust-debug [hunk 8]
```diff
@@ -1,6 +1,6 @@
 Stats {
     required_error: 0,
     matched_error: 0,
-    extra_error: 3,
+    extra_error: 6,
```
Comment (sunrabbit123) [inline crates/stc_ts_type_checker/tests/conformance/controlFlow/controlFlowBindingElement.ts.stats.rust-debug:4]: the problem is assign intersection

Comment (sunrabbit123) [reply crates/stc_ts_type_checker/tests/conformance/controlFlow/controlFlowBindingElement.ts.stats.rust-debug:4]: I think it's right to deal with it somewhere else.

File: crates/stc_ts_file_analyzer/src/analyzer/types/mod.rs [hunk 9]
```diff
@@ -927,13 +932,14 @@ impl Analyzer<'_, '_> {
         }
 
         let is_symbol = normalized_types.iter().any(|ty| ty.is_symbol());
-        let is_str = normalized_types.iter().any(|ty| ty.is_str());
+        let is_str = normalized_types.iter().any(|ty| ty.is_str() || ty.is_str_lit());
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/types/mod.rs:935]: Please revert this line. `is_str`is not a check for `string`. It's a check for string or string literal

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/stc_ts_file_analyzer/src/analyzer/types/mod.rs (hunk 9 above)
Comment (sunrabbit123) [reply crates/stc_ts_file_analyzer/src/analyzer/types/mod.rs:935]: but, string & "asdf" is string, number & "asdf" is never
It is right action in intersection

Comment (kdy1) [reply crates/stc_ts_file_analyzer/src/analyzer/types/mod.rs:935]: I mean `ty.is_str() || ty.is_str_lit()` is same as `ty.is_str()`.
Look at the code of function.

Comment (sunrabbit123) [reply crates/stc_ts_file_analyzer/src/analyzer/types/mod.rs:935]: I fixed it

File: crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs [hunk 10]
```diff
@@ -1048,7 +1051,46 @@ impl Analyzer<'_, '_> {
...
+                return Ok(Some(self.normalize(Some(span), Cow::Owned(ty), Default::default())?.into_owned()));
+            }
+        }
+        let result = match type_mode {
+            TypeOfMode::LValue => {
+                let mut temp_vec = vec![];
+                for elem in res_vec {
+                    if let Type::Intersection(ty) = elem.normalize() {
+                        for item in ty.types.iter() {
+                            temp_vec.push(item.normalize().clone());
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:1073]: `normalize().clone()` is an anti-pattern.

File: crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs [hunk 11]
```diff
@@ -955,6 +955,9 @@ impl Analyzer<'_, '_> {
             if read_only_flag {
                 return Err(ErrorKind::ReadOnly { span }.into());
             }
+            if let Some(ty) = matching_elements.pop() {
+                return Ok(Some(self.normalize(Some(span), Cow::Owned(ty), Default::default())?.into_owned()));
+            }
             return Ok(matching_elements.pop());
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:961]: Please change this line to `unreachable!()`

File: crates/stc_ts_file_analyzer/src/analyzer/types/mod.rs [hunk 12]
```diff
@@ -893,6 +894,9 @@ impl Analyzer<'_, '_> {
                         kind: TsKeywordTypeKind::TsUnknownKeyword,
                         ..
                     }) => {}
+                    Type::Param(TypeParam { constraint: Some(ty), .. }) => {
+                        normalized_types.push(ty.normalize().clone());
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/types/mod.rs:898]: Same here

File: crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs (hunk 10 above)
Comment (sunrabbit123) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:1073]: So how do you take ownership and put it in a vector while melting freezed?

Comment (kdy1) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:1073]: You don't need to remove `freezed` state, and you **should never** try to store freezed type as unfreezed form

Comment (sunrabbit123) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:1073]: How do you deup if there is a mix of friends saved as frezzled and friends who are not?

Comment (kdy1) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:1073]: `dedup_type` works with freezed types without any problem

Comment (sunrabbit123) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs]: oh.. thx I understand
I fixed their

File: crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs [hunk 13]
```diff
@@ -955,7 +955,10 @@ impl Analyzer<'_, '_> {
             if read_only_flag {
                 return Err(ErrorKind::ReadOnly { span }.into());
             }
-            return Ok(matching_elements.pop());
+            if let Some(ty) = matching_elements.pop() {
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:958]: Is this change necessary?

Comment (kdy1) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:958]: Basically, `Analyzer.normalize` should called only if required

Comment (kdy1) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:958]: I think you should patch usage side instead of return value

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs [hunk 14]
```diff
@@ -1048,7 +1051,46 @@ impl Analyzer<'_, '_> {
...
+
+        for el in matching_elements.into_iter() {
+            if let Ok(res) = self.normalize(Some(span), Cow::Owned(el), Default::default()) {
+                res_vec.push(res.into_owned());
+            }
+        }
+        res_vec.dedup_type();
+        if res_vec.len() == 1 {
+            if let Some(ty) = res_vec.pop() {
+                return Ok(Some(self.normalize(Some(span), Cow::Owned(ty), Default::default())?.into_owned()));
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:1064]: I think this doesn't seem right. We already normalized all types in `res_vec`

File: crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs [hunk 15]
```diff
@@ -1048,7 +1051,46 @@ impl Analyzer<'_, '_> {
...
+            }
+        }
+        let result = match type_mode {
+            TypeOfMode::LValue => {
+                let mut temp_vec = vec![];
+                for elem in res_vec {
+                    if let Type::Intersection(ty) = elem.normalize() {
+                        for item in ty.types.iter() {
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:1072]: Please use `.extend(ty.types.iter().cloned())` instead

File: crates/stc_ts_file_analyzer/tests/pass/types/mapped/recursiveMappedTypes/1.ts [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+  [P in keyof T]: undefined extends T[P] ? never : P;
+}[keyof T];
+type Child<T> = { [P in NonOptionalKeys<T>]: T[P] };
+
+export interface ListWidget {
+  type: "list";
+  minimum_count: number;
+  maximum_count: number;
+  collapsable?: boolean; //default to false, means all expanded
+  each: Child<ListWidget>;
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/tests/pass/types/mapped/recursiveMappedTypes/1.ts:13]: This test file should not stored in `pass`.

https://www.typescriptlang.org/play?#code/PTAECUFMAcCcHtQDMEFtQGIBMBOPWAoAgFwE9pJQA5eAOwHlpiBLOgQwBsBpSUgZwA8AFQB8oALygA3gVCgA2gAVQzWqADWveElBCAugC5QAV1oATSElWQzoSAA9ikc311K9oAPyhakAG6QsKBGigDcBAC+8pqk2rp64WQUoADCABbMHGbCYpJSCsqq1HSMLOzcvIKihm6KHhHhBA7Q8LDEKrROsEhsAMaUADLMfMQA6sxmAOaQ7TJySZBGAEQcw8RL4XKoqsyoxqgA+r3wpsRGtPsARoGboKhs9rv7Ryed51c3sqDHHBxs0Hw2JcOJBPEZLvB4CC2LRQqAQBYesYOO1iIgehw+JAADR3SAw1ycDh2ezQGEWMxffG9NJGdKZbJDEbjKYzEThCJEBagJnEelZCSpDJZAS8lnTYjsogWXp-WCUEHtexGXn8szhewAOgW4SAA

Comment (sunrabbit123) [reply crates/stc_ts_file_analyzer/tests/pass/types/mapped/recursiveMappedTypes/1.ts:13]: If you put this in the error box, you won't pass the test.
I don't have an error with the circulation reference.

Why don't you just ignore it?

Comment (sunrabbit123) [reply crates/stc_ts_file_analyzer/tests/pass/types/mapped/recursiveMappedTypes/1.ts:13]: I will append `.` at file name

Comment (kdy1) [reply crates/stc_ts_file_analyzer/tests/pass/types/mapped/recursiveMappedTypes/1.ts:13]: You should move it to `visualize` instead

Comment (sunrabbit123) [reply crates/stc_ts_file_analyzer/tests/pass/types/mapped/recursiveMappedTypes/1.ts:13]: what is mean `visualize test`?

Comment (kdy1) [reply crates/stc_ts_file_analyzer/tests/pass/types/mapped/recursiveMappedTypes/1.ts:13]: file analyzer/tests/visualize

Comment (sunrabbit123) [reply crates/stc_ts_file_analyzer/tests/pass/types/mapped/recursiveMappedTypes/1.ts:13]: I was just wondering,
What does the test mean?

Comment (kdy1) [reply crates/stc_ts_file_analyzer/tests/pass/types/mapped/recursiveMappedTypes/1.ts]: https://github.com/dudykr/stc/tree/d3d6a8fa98f5d6cceefb046124fcf6b63d9305cb/crates/stc_ts_file_analyzer#tests

Comment (sunrabbit123) [reply crates/stc_ts_file_analyzer/tests/pass/types/mapped/recursiveMappedTypes/1.ts]: My vision was dark. I'm so sorry 
I fixed its

File: crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs [hunk 17]
```diff
@@ -1048,7 +1048,38 @@ impl Analyzer<'_, '_> {
...
+                }
+                temp_vec.dedup_type();
+                Type::Intersection(Intersection {
+                    span,
+                    types: temp_vec,
+                    metadata: Default::default(),
+                    tracker: Default::default(),
+                })
+            }
+            TypeOfMode::RValue => Type::union(res_vec),
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:1080]: Please use `Type::new_union` instead

File: crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs [hunk 18]
```diff
@@ -1048,7 +1048,38 @@ impl Analyzer<'_, '_> {
...
+                res_vec.push(res.into_owned());
+            }
+        }
+        res_vec.dedup_type();
+        if res_vec.len() == 1 {
+            return Ok(res_vec.pop());
+        }
+        let result = match type_mode {
+            TypeOfMode::LValue => {
+                let mut temp_vec = vec![];
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:1064]: I think `Type::new_intersection(span, res_vec)` will do what you are trying to do

Comment (kdy1) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:1064]: And more efficient

File: crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs [hunk 19]
```diff
@@ -2452,12 +2484,29 @@ impl Analyzer<'_, '_> {
                             ..opts
                         },
                     ) {
-                        Ok(ty) => tys.push(ty),
+                        Ok(ty) => {
+                            if let Type::Union(ty::Union { mut types, .. }) = ty {
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:2488]: I think we need normalization here.

Comment (kdy1) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:2488]:
```rust

if ty.is_union_type() {
    tys.extend(ty.expect_union_type().types);
    continue;
}

```


will work

**Review by kdy1 — CHANGES_REQUESTED** [review]

**Comment by sunrabbit123** [comment]
I fixed their

File: crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs (hunk 18 above)
Comment (kdy1) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:1064]: Seems like this is not fixed

Comment (kdy1) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:1064]: I meant 1064 ~ 1072 can be replaced as a whole.
`new_intersection` has logic for 

```rust
                for elem in res_vec {
                    if let Type::Intersection(ty) = elem.normalize() {
                        temp_vec.extend(ty.types.iter().cloned());
                    } else {
                        temp_vec.push(elem);
                    }
                }
                temp_vec.dedup_type();
```

Comment (sunrabbit123) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs:1064]: Oh, the next code was in new_intersection
```rs
for ty in iter {
    if ty.is_intersection() {
        tys.extend(ty.expect_intersection().types);
    } else {
        tys.push(ty);
    }
}
tys.dedup_type();
```

I will replace it

thx!

Comment (sunrabbit123) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs]: I fixed it

**Review by kdy1 — APPROVED** [review]

**Comment by kdy1** [comment]
Thank you!
